### ISSUE TITLE
net.imglib2.util.RealSum reveals bug in Java 8 (1.8.0_25 and 1.8.0_40) on Mac OS-X 

### ISSUE THREAD

**Issue by MichaelEllis** [imglib/imglib2#79]
I have found the unit test RealSumTest.java fails on testAdd() when run under Java 8 (1.8.0_25 and 1.8.0_40). I believe Java 8 is not officially supported by ImgLib2 but I though perhaps forewarned is forearmed.

After some digging I found the RealSum::add() function appears to expose a bug in Java 8. (at least on my OS-X 10.9.4 / MacBook Retina)

The code at present is as follows:
```
    /**
     * Add an element to the sum. All intermediate sums are updated and the
     * capacity is increased on demand.
     *
     * @param a the summand to be added
     */
    final public void add(final double a) {
        int i = 0;
        double s = a;
        try {
            while (flags[i]) {
                flags[i] = false;
                s += sums[i];
                sums[i] = 0.0;
                ++i;
            }
            flags[i] = true;
            sums[i] = s;
            return;
        } catch (final IndexOutOfBoundsException e) {
            expand(s);
        }
    }
```

The problem seems to be that the value of `s` is corrupted by the time the `IndexOutOfBoundsException` is caught. This can be shown with the following code:

```
    final public void addIllustrateBug(final double a) {
        int i = 0;
        double s = a;
        try {
            while (flags[i]) {
                flags[i] = false;
                s += sums[i];
                sums[i] = 0.0;
                ++i;
                if (i >= flags.length) {
                    System.out.printf("about to go bust with %g%n", s);
                }
            }
            flags[i] = true;
            sums[i] = s;
            return;
        } catch (final IndexOutOfBoundsException e) {
            System.out.printf("expanding with %g%n", s);
            expand(s);
        }
    }
``` 
If and when this happens seems to be unpredictable but I enclose output from my log here:

```
about to go bust with 2.00000
expanding with 2.00000
expand(2.00000)
about to go bust with 4.00000
expanding with 4.00000
expand(4.00000)
about to go bust with 8.00000
expanding with 8.00000
expand(8.00000)
about to go bust with 16.0000
expanding with 16.0000
expand(16.0000)
about to go bust with 32.0000
expanding with 32.0000
expand(32.0000)
about to go bust with 64.0000
expanding with 64.0000
expand(64.0000)
about to go bust with 128.000
expanding with 128.000
expand(128.000)
about to go bust with 256.000
expanding with 256.000
expand(256.000)
about to go bust with 512.000
expanding with 512.000
expand(512.000)
about to go bust with 1024.00
expanding with 1024.00
expand(1024.00)
about to go bust with 2048.00
expanding with 1.47358e-61
expand(1.47358e-61)
about to go bust with 2048.00
expanding with 1.47358e-61
expand(1.47358e-61)
.
.
.
```

As you can see, things go pear-shaped at "about to go bust with 2048.00"!

Rewriting `add()` to explicitly test for out of bounds restores peace, harmony and wellbeing (although I do not know how this hits performance).

```
    final public void addFixed(final double a) {
        int i = 0;
        double s = a;
        while (flags[i]) {
            flags[i] = false;
            s += sums[i];
            sums[i] = 0.0;
            ++i;
            if (i >= flags.length) {
                System.out.printf("about to go bust with %g%n", s);
                expand(s);
                break;
            }
        }
        flags[i] = true;
        sums[i] = s;
        return;
    }
```

Some of the ImgLib2 Examples fail under Java 8; Example6c (the one that uses FFT to find the position of a sub image in a larger image). I wonder if some of the examples could be wrapped up in JUnit tests to check that they were always delivering the goods?

**Comment by axtimwalde** [comment]
Hi @MichaelEllis sorry for being resistant (or ignorant) to accept the workaround.  I just looked at your [Java bug report] (https://bugs.openjdk.java.net/browse/JDK-8066900?page=com.atlassian.jira.plugin.system.issuetabpanels:changehistory-tabpanel) and it looks like this got fixed in hotspot.  So may be we were lucky.  Can you confirm?

**Comment by MichaelEllis** [comment]
Hi @axtimwalde, Yes I can confirm that bug fixed at java version "1.8.0_45"

So good news all around.

**Comment by axtimwalde** [comment]
Glad to hear that!  Closing issue as "Fixed by who caused it.".
